src/t8_IO/t8_IO_cxx.hxx

```diff
@@ -0,0 +1,195 @@
+*/
+
+/**
+ * \file t8_IO_cxx.hxx
+ * This file describes base classes for Input and output routines.
+ * For every supported input/output type a reader/writer has to be implemented
+ * in t8_reader/t8_writer.
+ */
+#ifndef T8_IO_CXX_HXX
+#define T8_IO_CXX_HXX
+
+#include <src/t8_IO/t8_IO.h>
+#include <sc_mpi.h>
+
+T8_EXTERN_C_BEGIN ();
+
+typedef void        t8_extern_t;
+
+typedef enum t8_partition
+{
+  T8_NO_PARTITION = 0,
+  T8_PARTITION
+} t8_partition_t;
+
+typedef enum t8_read_status
+{
+  T8_READ_SUCCESS = 0,
+  T8_READ_FAIL
+} t8_read_status_t;
+
+typedef enum t8_write_status
+{
+  T8_WRITE_SUCCESS = 0,
+  T8_WRITE_FAIL
+} t8_write_status_t;
+
+/**
+ * Base-Class for reader-routines.
+ */
+typedef struct t8_IO_reader
+{
+private:
+  /**
+   * The communicator used by the reader.
+   */
+  sc_MPI_Comm comm = sc_MPI_COMM_WORLD;
+  /**
+   * The process that coordinates the reading.
+   * If a file should only be read unpartitioned, this is the
+   * process that reads the file.
+   */
+  unsigned int        main_proc = 0;
+
+  t8_partition_t      partition = T8_NO_PARTITION;
+public:
+  /** The destructor. It does nothing but has to be defined since
+ * we may want to delete an t8_IO_reader that is actually inherited
+ * and providing an implementation for the destructor ensures that the
+ * destructor of the child class will be executed. */
+                      virtual ~ t8_IO_reader ()
+  {
+  };
+
+    /**
+   * A reader function, that translates an external object into a forest.
+   */
+  virtual t8_read_status_t read (void) = 0;
+  /**
+   *  Set the Communicator
+   * 
+   * \param new_comm The new communicator to use.
+   */
+  void                set_Communicator (sc_MPI_Comm new_comm)
+  {
+    comm = new_comm;
+  }
+
+  /**
+   * Get the Communicator
+   * 
+   * \return sc_MPI_Comm 
+   */
+  sc_MPI_Comm         get_Communicator ()
+  {
+    return comm;
+  }
+
+  /**
+   * Set the main proc 
+   * 
+   * \param proc The rank of the process to use
+   */
+  void                set_main_proc (const unsigned int proc)
+  {
+    main_proc = proc;
+  }
+
+  /**
+   * Get the main proc
+   * 
+   * \return unsigned int The main-reader proc.
+   */
+  unsigned int        get_main_proc ()
+  {
+    return main_proc;
+  }
+
+  /**
+   * Set partition 
+   * 
+   * \param part Flag, if partitioning should be used or not.
+   */
+  void                set_partition (const t8_partition_t part)
+  {
+    partition = part;
+  }
+
+  /**
+   * Get the partition-flag
+   * 
+   * \return t8_partition_t The partition-flag.
+   */
+  t8_partition_t      get_partition ()
+  {
+    return partition;
+  }
+  /**
+   * Set the source object, if source is not NULL
+   * 
+   * \param[in] source an object to be filled.
+   * \return t8_write_status_t T8_WRITE_FAIL if it wasn't able to set the source, T8_WRITE_SUCCESS otherwise
+   */
+  virtual t8_read_status_t set_source (const t8_extern_t * source) = 0;
```

Comment:
This routine does not have be part of the base class. We should talk about this.